lib/srv/desktop/rdp/rdpclient/src/rdpdr/scard.rs

```diff
@@ -2172,20 +2348,30 @@ impl Contexts {
         ctx
     }
 
-    fn get(&mut self, id: u32) -> Option<&mut ContextInternal> {
-        self.contexts.get_mut(&id)
+    fn get(&mut self, id: u32) -> RdpResult<&mut ContextInternal> {
+        self.contexts
+            .get_mut(&id)
+            .ok_or_else(|| invalid_data_error(&format!("unknown context id: {}", id)))
     }
 
     fn release(&mut self, id: u32) {
         self.contexts.remove(&id);
     }
 }
 
-#[derive(Debug, PartialEq)]
+#[derive(Debug)]
 struct ContextInternal {
     handles: HashMap<u32, piv::Card<TRANSMIT_DATA_LIMIT>>,
     next_id: u32,
     cache: HashMap<String, Vec<u8>>,
+    // If we receive a GetStatusChange_Call with an infinite timeout, we need to
+    // return a GetStatusChange_Return (embedded in a DeviceControlResponse) with
+    // its return code set to SCARD_E_CANCELLED in the case that we receive a
+    // SCARD_IOCTL_CANCEL.
+    //
+    // This value will be set during the handling of the GetStatusChange_Call, so that
+    // it can be fetched and returned in response to a SCARD_IOCTL_CANCEL.
+    scard_cancel_response: Option<DeviceControlResponse>,
```

Comment:
We somehow need to represent the fact that this context has a received a `SCARD_IOCTL_GETSTATUSCHANGEW` previously (which can be thought of like an RPC) that's hanging, and which we may return from at later point if we get an `SCARD_IOCTL_CANCEL`. We need to at least hold on to that `SCARD_IOCTL_GETSTATUSCHANGEW`'s `completion_id` in the context, in which case I would imagine implementing that like `scard_cancel_completion_id: Option<u32>`. (Whether that design is preferable per separation of concerns is debatable, but beside the point).

Similar to your other comment, I don't share the feeling of weirdness, but am open to suggestions for how it can be improved. I can come up with different ways to do this, but none of them appear obviously better.

Reply:
What if the context held a vec of responses that should be sent back on the next `SCARD_IOCTL_CANCEL`?